client/src/main/java/io/cloudsoft/winrm4j/client/WinRmClient.java

```diff
@@ -328,11 +329,18 @@ public ReceiveResponse call() {
 
     private void assertFaultCode(SOAPFaultException soapFault, String code) {
         try {
-            if (soapFault.getFault().getDetail().getFirstChild().getAttributes().getNamedItem("Code").getNodeValue().equals(code)) {
-                LOG.trace("winrm client {} received error 500 response with code {}, response {}", this, code, soapFault);
-            } else {
-                throw soapFault;
+            NodeList faultDetails = soapFault.getFault().getDetail().getChildNodes();
+            for (int i = 0; i < faultDetails.getLength(); i++) {
+                if (faultDetails.item(i).getLocalName().equals("WSManFault")) {
+                    if (faultDetails.item(i).getAttributes().getNamedItem("Code").getNodeValue().equals(code)) {
+                        LOG.trace("winrm client {} received error 500 response with code {}, response {}", this, code, soapFault);
+                        return;
+                    } else {
+                        throw soapFault;
+                    }
+                }
             }
+            throw soapFault;
         } catch (NullPointerException e) {
```

Comment:
Try try-catch approach for handling nulls isn't good when you're looping over multiple faults. We might fail on the first fault with an NPE, even though the second fault was the one with the matching code.

Also do we really want to `throw soapFault` in the if block, within the for loop? Or do we want to wait until the end to see if any of the `WSManFault`s had the matching code?

I guess you've added the for loop because of seeing a specific situation of a WSManFault along with another child fault. I can see how your code changes would solve that specific situation, but I presume we can generalise it further.

If we ever get given multiple `WSManFault` instances, it's unclear what the best behaviour is: should it log.trace + return if any of them match the given code, or should it throw if there is one that does not match the code?